psalm integration

snapshotpl opened the pull request:
|    Q          |   A
|-------------- | ------
| QA            | yes

### Description

Closes #53

snapshotpl committed:
psalm integration

Signed-off-by: Witold Wasiczko <witold@wasiczko.pl>

snapshotpl committed:
Add baseline

Signed-off-by: Witold Wasiczko <witold@wasiczko.pl>

snapshotpl committed:
Add baseline

Signed-off-by: Witold Wasiczko <witold@wasiczko.pl>

snapshotpl committed:
Fix some issues found by psalm

Signed-off-by: Witold Wasiczko <witold@wasiczko.pl>

snapshotpl committed:
Fix some issues found by psalm

Signed-off-by: Witold Wasiczko <witold@wasiczko.pl>

snapshotpl committed:
Fix type

Signed-off-by: Witold Wasiczko <witold@wasiczko.pl>

snapshotpl committed:
Fix type

Signed-off-by: Witold Wasiczko <witold@wasiczko.pl>

boesing commented on src/MessageTrait.php:
```diff
@@ -37,13 +37,17 @@ trait MessageTrait
      *
      * @var array
+     *
+     * @psalm-var array<array-key, array<array-key, string>>
```
```suggestion
     * @psalm-var array<non-empty-string, list<string>>
```

boesing commented on src/MessageTrait.php:
```diff
@@ -37,13 +37,17 @@ trait MessageTrait
      *
      * @var array
+     *
+     * @psalm-var array<string, string>
```
```suggestion
     * @psalm-var array<non-empty-string, non-empty-string>
```

boesing commented on src/MessageTrait.php:
```diff
@@ -110,6 +114,8 @@ public function withProtocolVersion($version) : MessageInterface
      * @return array Returns an associative array of the message's headers. Each
      *     key MUST be a header name, and each value MUST be an array of strings.
+     *
+     * @psalm-return array<array-key, array<array-key, string>>
```
```suggestion
     * @psalm-return array<non-empty-string, list<string>>
```

boesing commented on src/Response.php:
```diff
@@ -37,6 +37,8 @@ class Response implements ResponseInterface
      *
      * @var array
+     *
+     * @psalm-var array<int, string>
```
```suggestion
     * @psalm-var array<positive-int, non-empty-string>
```

boesing reviewed (changes requested):
I prefer having `@var array<x,y>` annotations over `@psalm-var` as these are valid in most IDEs as of now.
Afaik, `list<x>` is still not supported by most of the IDEs.

I know that @Ocramius has some opinions on that.

In my company, we are adding:

```php
/**
 * @var array<int,string>
 * @psalm-var list<string>
 */
```

Ocramius commented:
> I prefer having `@var array<x,y>` annotations over `@psalm-var` as these are valid in most IDEs as of now.
Afaik, `list<x>` is still not supported by most of the IDEs.

If you want IDE support, you can add `@var` and `@psalm-var` in the same block.

Careful with removing `list<T>`! `list<T>` is a much stricter (better) type than `array<int, T>`, so if you really want IDE support, do not remove it, but add a `@var` block

Ocramius commented on psalm.xml.dist:
```diff
@@ -0,0 +1,33 @@
+        <InternalMethod>
+            <errorLevel type="suppress">
+                <referencedMethod name="PHPUnit\Framework\MockObject\Builder\InvocationMocker::method"/>
+            </errorLevel>
```
I don't remember if this is copied from somewhere, but these `<errorLevel` blocks can be merged together 👍

Ocramius replied:
Also: not a blocker 👍

Ocramius reviewed (approved):
I think @boesing's suggestions cover most of what needs to be improved at type level: once that's through, 🚢

weierophinney commented:
@snapshotpl I've rebased against current 2.6.x, which includes the cutover to GHA. The Psalm check is failing, and it looks like there's still some feedback that needs to be addressed; ping me when done, and I'll do a final review.

lcobucci commented:
> Afaik, list<x> is still not supported by most of the IDEs.

Everything works on PHPStorm 2021.1 here (both on `@var` and `@param`) :+1:

snapshotpl committed:
Fix psalm issues

Signed-off-by: Witold Wasiczko <witold@wasiczko.pl>

snapshotpl committed:
Apply suggestions from code review

Co-authored-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>
Signed-off-by: Witold Wasiczko <witold@wasiczko.pl>

snapshotpl committed:
Improve psalm config

Signed-off-by: Witold Wasiczko <witold@wasiczko.pl>

snapshotpl committed:
Update baseline

Signed-off-by: Witold Wasiczko <witold@wasiczko.pl>

snapshotpl commented:
@weierophinney done

boesing commented on src/AbstractSerializer.php:
```diff
@@ -141,7 +143,7 @@ protected static function serializeHeaders(array $headers) : string
      * Filter a header name to wordcase
      */
-    protected static function filterHeader($header) : string
+    protected static function filterHeader(string $header) : string
```
Adding native type-hint to non-private methods on a non-final class is a BC break due to mismatching signature.

boesing commented on src/CallbackStream.php:
```diff
@@ -104,7 +104,7 @@ public function isSeekable() : bool
      * {@inheritdoc}
      */
-    public function seek($offset, $whence = SEEK_SET)
+    public function seek($offset, $whence = SEEK_SET) : void
```
class is not final, thus this is a BC break

boesing commented on src/MessageTrait.php:
```diff
@@ -138,20 +144,20 @@ public function hasHeader($header) : bool
      * empty array.
      *
-     * @param string $header Case-insensitive header field name.
+     * @param string $name Case-insensitive header field name.
```
Is there a specific reason why you would rename this? Just curious tho.
I am pretty sure that this might be a BC break aswell, as we already have PHP 8 support which added method calls with named arguments which will break if you change this.

But as we are implementing psalm, I'd prefer having the `@no-named-arguments` in here. Maybe we move this change to the next major even tho I don't understand why this change is here in the first place.

boesing commented on src/Request.php:
```diff
@@ -57,10 +57,10 @@ public function getHeaders() : array
      * {@inheritdoc}
      */
-    public function getHeader($header) : array
+    public function getHeader($name) : array
```
Is there a specific reason why you would rename this? Just curious tho.
I am pretty sure that this might be a BC break aswell, as we already have PHP 8 support which added method calls with named arguments which will break if you change this.

But as we are implementing psalm, I'd prefer having the `@no-named-arguments` in here. Maybe we move this change to the next major even tho I don't understand why this change is here in the first place.

weierophinney committed:
qa: incorporate feedback from @boesing

- Revert addition of scalar type hint in `AbstractSerializer::serializeHeaders` to eliminate BC break
- Revert addition of void return type in `CallbackStream::seek` to eliminate BC break
- Revert parameter name change in `MessageTrait::hasHeader()` to eliminate BC break

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: add parameter type hint annotations to CallbackStream::seek method

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: update phpunit config to latest schema

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: revert variable name change

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
